pic-sure-auth-services/src/main/java/edu/harvard/hms/dbmi/avillach/auth/service/impl/AuthenticationService.java

```diff
@@ -0,0 +1,163 @@
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import edu.harvard.hms.dbmi.avillach.auth.entity.Connection;
+import edu.harvard.hms.dbmi.avillach.auth.exceptions.NotAuthorizedException;
+import edu.harvard.hms.dbmi.avillach.auth.model.response.PICSUREResponse;
+import edu.harvard.hms.dbmi.avillach.auth.repository.ConnectionRepository;
+import edu.harvard.hms.dbmi.avillach.auth.utils.RestClientUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+import edu.harvard.hms.dbmi.avillach.auth.entity.User;
+import edu.harvard.hms.dbmi.avillach.auth.repository.UserRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.http.client.ClientHttpRequestFactory;
+import org.springframework.http.client.SimpleClientHttpRequestFactory;
+import org.springframework.stereotype.Service;
+
+/**
+ * This class provides authentication functionality. This implements an authenticationService interface
+ * in the future to support different modes of authentication.
+ *
+ * <h3>Thoughts of design</h3>
+ * The main purpose of this class is returns a token that includes information of the roles of users.
+ */
+
+@Service
+public class AuthenticationService {
+
+    private final Logger logger = LoggerFactory.getLogger(AuthenticationService.class);
+
+    private final OauthUserMatchingService matchingService;
+
+    private final UserRepository userRepository;
+
+    private final BasicMailService basicMailService;
+
+    private final UserService userService;
+    private static final int AUTH_RETRY_LIMIT = 3;
+
+    private String deniedEmailEnabled;
+
+    private final String auth0host;
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
+    private final ConnectionRepository connectionRepository;
+    private final RestClientUtil restClientUtil;
+
+    @Autowired
+    public AuthenticationService(OauthUserMatchingService matchingService, UserRepository userRepository, BasicMailService basicMailService, UserService userService,
+                                 @Value("${application.denied.email.enabled}") String deniedEmailEnabled, @Value("${application.auth0.host}") String auth0host, ConnectionRepository connectionRepository, RestClientUtil restClientUtil) {
+        this.matchingService = matchingService;
+        this.userRepository = userRepository;
+        this.basicMailService = basicMailService;
+        this.userService = userService;
+        this.deniedEmailEnabled = deniedEmailEnabled;
+        this.auth0host = auth0host;
+        this.connectionRepository = connectionRepository;
+        this.restClientUtil = restClientUtil;
+    }
+
+    public ResponseEntity<?> getToken(Map<String, String> authRequest) throws IOException {
```

Comment:
don't like seeing this class in the service layer. I'd like the controller to take the response from the service layer and determine stuff like the http code in the controller itself.

Reply:
I absolutely agree. I refactored a ton of theses. I will update this.